File: bk2ch13p635activityView/ch26p901activityView/MyElaborateActivity.m

#import "MyElaborateActivity.h"
#import "ElaborateViewController.h"

@interface MyElaborateActivity()
@property (nonatomic, strong) NSArray* items;
@property (nonatomic, strong) UIImage* image;
@end

// in iOS 7, this works on iPhone and on iPad


@implementation MyElaborateActivity

-(id)init {
    self = [super init];
    if (self) {
        CGFloat scale = (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) ? 76 : 60;
        scale -= 10;
        UIImage* im = [UIImage imageNamed:@"sunglasses.png"];
        CGFloat largerSize = fmaxf(im.size.height, im.size.width);
        scale /= largerSize;
        CGSize sz = CGSizeMake(im.size.width*scale, im.size.height*scale);
        UIGraphicsBeginImageContextWithOptions(sz, NO, 0);
        [im drawInRect: (CGRect) {CGPointZero, sz}];
        im = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();
        self->_image = im;
    }
    return self;
}

+(UIActivityCategory)activityCategory {
    return UIActivityCategoryAction;
}

-(NSString *)activityType {
    return @"com.neuburg.matt.elaborateActivity";
}

-(NSString *)activityTitle {
    return @"Elaborate";
}

-(UIImage *)activityImage {
    return self.image;
}

-(BOOL)canPerformWithActivityItems:(NSArray *)activityItems {
    NSLog(@"can perform %@", activityItems);
    return YES;
}

-(void)prepareWithActivityItems:(NSArray *)activityItems {
    NSLog(@"prepare %@", activityItems);
    self.items = activityItems;
}

-(UIViewController *)activityViewController {
    ElaborateViewController* evc = [ElaborateViewController new];
    evc.activity = self;
    evc.items = self.items;
    return evc;
}

-(void) dealloc {
    NSLog(@"%@", @"elaborate activity dealloc");
}



@end

File: bk2ch13p635activityView/ch26p901activityView/MyElaborateActivity.swift

import UIKit

class MyElaborateActivity : UIActivity {
    
    var items : [Any]?
    var image : UIImage
    
    override init() {
        let idiom = UIScreen.main.traitCollection.userInterfaceIdiom
        var scale : CGFloat = (idiom == .pad ? 76 : 60) - 10
        let im = UIImage(named:"sunglasses.png")!
        let largerSize = fmax(im.size.height, im.size.width)
        scale /= largerSize
        let sz = CGSize(im.size.width*scale, im.size.height*scale)
        let r = UIGraphicsImageRenderer(size:sz, format:im.imageRendererFormat)
        self.image = r.image { _ in
            im.draw(in:CGRect(origin: .zero, size: sz))
        }
        super.init()
    }
    
    override class var activityCategory : UIActivity.Category {
        return .action // the default
    }
    
    override var activityType : UIActivity.ActivityType? {
        return UIActivity.ActivityType("com.neuburg.matt.elaborateActivity")
    }
    
    override var activityTitle : String? {
        return "Elaborate"
    }
    
    override var activityImage : UIImage? {
        return self.image
    }
    
    override func canPerform(withActivityItems activityItems: [Any]) -> Bool {
        print("elaborate can perform \(activityItems)")
        print("returning true")
        return true
    }
    
    override func prepare(withActivityItems activityItems: [Any]) {
        print("elaborate prepare \(activityItems)")
        self.items = activityItems
    }
    
    override var activityViewController : UIViewController? {
        let mvc = MustacheViewController(activity: self, items: self.items!)
        return mvc
    }
    
    deinit {
        print("elaborate activity dealloc")
    }
    
}
